TreeMap impl based on AVL-tree

sergey-melnychuk opened the pull request:
TreeMap based on AVL-tree

| Method | Complexity (worst case) | 
|---------|-----------------------|
| `get`/`contains_key` | O(1) // UnorderedMap lookup |
| `insert`/`remove` | O(log(N)) |
| `min`/`max` | O(log(N)) |
| `above`/`below` | O(log(N)) |
| `range` of K elements | O(Klog(N)) |

Closes https://github.com/near/near-sdk-rs/issues/146

sergey-melnychuk commented:
@nearmax , please check out the draft (please note it is a draft: I'll catch up with all TODOs) - does it look like what's expected? 

`Heap` incapsulates indices, thus `HeapMap` only maps keys to values, letting `Heap` do all housekeeping of min-heap. In final version `insert` & `remove` will be O(log(N)), and `iterator` will be O(Nlog(N)), as promised.

Half of CI checks fail (for `env: NEAR_RELEASE=false`) with error `HostError(GasLimitExceeded)`, am I doing something wrong? :)

nearmax commented:
Hey @sergey-melnychuk ! Thank you for working on it. Sorry for not reviewing it yesterday, I will review it today.

sergey-melnychuk commented:
@nearmax , no problem, PR is ready for review now.

nearmax commented on near-sdk/src/collections/heap_map.rs:
```diff
@@ -0,0 +1,183 @@
+        K: Ord + BorshSerialize + BorshDeserialize,
+        V: BorshSerialize + BorshDeserialize,
+{
+    pub fn new(id: Vec<u8>) -> Self {
```
Please add method that exposes the largest key (the top element from the heap).

sergey-melnychuk replied:
done

nearmax commented on near-sdk/src/collections/heap_map.rs:
```diff
@@ -0,0 +1,183 @@
+    }
+
+    // Note: mutable borrow is required for iterator.
+    pub fn entries<'a>(&'a mut self) -> impl Iterator<Item = (K, V)> + 'a {
```
Please rename to `sort_iter`

nearmax commented on near-sdk/src/collections/heap_map.rs:
```diff
@@ -0,0 +1,183 @@
+    }
+
+    // Note: iterator mutates underlying indexes when sorts the keys.
+    pub fn keys<'a>(&'a mut self) -> impl Iterator<Item = K> + 'a {
```
Please rename to `sort_keys`

nearmax commented on near-sdk/src/collections/heap.rs:
```diff
@@ -0,0 +1,392 @@
+        self.indices.remove(&value);
+    }
+
+    pub fn iter<'a>(&'a mut self) -> impl Iterator<Item = T> + 'a {
```
Please rename to `sort_iter`.

nearmax commented on near-sdk/src/collections/heap.rs:
```diff
@@ -0,0 +1,392 @@
+        self.elements.iter()
+    }
+}
+
```
Please add implementation of `Default` trait like in `UnorderedMap`.

nearmax commented on near-sdk/src/collections/heap.rs:
```diff
@@ -0,0 +1,392 @@
+        self.indices.clear();
+    }
+
+    pub fn at(&self, idx: u64) -> Option<T> {
```
Please add docstrings explaining that index is 1-based. Same for `lookup`.

nearmax commented on near-sdk/src/collections/heap_map.rs:
```diff
@@ -0,0 +1,183 @@
+        opt
+    }
+
+    // Note: iterator mutates underlying indexes when sorts the keys.
```
Please use `///`

nearmax commented:
My follow up request was to implement BTree on the key-value storage. This issue does not close entirely #146 but for the sake of gitcoin management we can mark it as such. HeapMap does not solve the problem, since users want to iterate over the collection with O(log n) or amortized O(1) per step. However, I see value in Heap and HeapMap, because someone might find it useful to have access in the top-most element (e.g. for the cases when there are order books, but not when entire order book needs to be retrieved).

nearmax commented:
@sergey-melnychuk 
> Half of CI checks fail (for env: NEAR_RELEASE=false) with error HostError(GasLimitExceeded), am I doing something wrong? :)

Your code is fine. We broke our CI recently: https://github.com/near/near-sdk-rs/issues/156

SkidanovAlex commented:
`Heap::insert` doesn't check if the key is already present, and `HeapMap::insert` doesn't do this check either. Thus, if the same key is inserted twice, it will be twice in the `elements` and once in the `indices`. If it is then removed, it will be once in `elements` and zero times in `indices`. At this point looking up by key will report the key missing, but iterating will report it present, i.e. the map will be inconsistent.

nearmax commented:
@sergey-melnychuk We will be happy to pay out the bounty, given your code matches your proposal that we accepted. Unfortunately, we haven't realized at the time of acceptance that the heap is not a suitable data structure for the ordered map, and will not be merging the implementation in, seeking further to have a sorted map implementation.
Alternatively, we will also be happy to pay extra $250 ($750 total) if you want to implement the proper sorted map, based on AVL, RedBlack, or some other sort of self-balancing trees.
Also note that your Heap implementation was not covering a rather trivial corner case. Should you decide to implement the balanced tree, we will expect higher attention to details

sergey-melnychuk commented:
@nearmax , @SkidanovAlex , thank you for the review, I will look into it today in the evening.

This is the initial approach, prototype, so work only starts here, not ends :) I was trying to prototype quickly - to get feedback and start for next iteration(s).

To summarize:
- double-insertion bug - my bad, thanks for pointing out, I will fix it today (fix/test must be simple).
- requirements: now desired features are more or less defined, so I can address them:
  - O(Klog(N)) for first K keys: I believe it is possible, heap seems like a good fit.
  - `floor`/`ceil` queries - unexpected in current impl, thus O(Nlog(N)).
- O(Nlog(N)) for sorted view without the heap is totally valid point, any array/vector would do.

Next impl must fit these:
- `min`/`max` queries in O(log(N)) worst case
- `floor`/`ceil` (find closes key above/below) queries in O(log(N)) worst case
- `insert`/`remove`/`lookup` in O(log(N)) worst case
- iterate keys in sorted order in O(Nlog(N)) worst case

IMHO AVL is the best fit in performance/complexity tradeoff. Let me come up with a prototype of AVL based on Maps.

nearmax commented:
@sergey-melnychuk 

> IMHO AVL is the best fit in performance/complexity tradeoff. Let me come up with a prototype of AVL based on Maps.

Thank you, AVL would be the best. And sorry for the mess up.

sergey-melnychuk commented:
@nearmax , nothing to be sorry about - it is much easier to clarify requirements in front of a prototype, than out of thin air :)

Added `TreeMap` skeleton, I believe in a few days I will make it ready for review. It looks promising, most operations seem to fit O(log(N)).

SkidanovAlex commented on near-sdk/src/collections/heap.rs:
```diff
@@ -0,0 +1,464 @@
+            assert_eq!(heap.len() as usize, n - 1 - i);
+        }
+
+        assert_eq!(heap.len(), 0);
```
Add an extra remove_maxp to test removing max from an empty heap.

sergey-melnychuk replied:
Nice catch, thanks.

sergey-melnychuk commented:
@nearmax , @SkidanovAlex PR is ready for review. Access/mutation takes O(log(N)) in worst case and iteration over K sorted elements takes O(Klog(N)) in worst case, as balance is restored in O(log(N)) after insertions and deletions. Now I see I should have started with AVL-tree from the beginning :) 

Though I was trying to keep code as clean as readable as possible, still added some comments to tricky parts: mostly related to removing a node, balancing and rotations.

Also HeapMap is removed, it was totally useless.

SkidanovAlex commented on near-sdk/src/collections/tree_map.rs:
```diff
@@ -0,0 +1,1300 @@
+        self.key.get(&at)
+    }
+
+    fn floor_at(&self, mut at: u64, key: &K) -> Option<K> {
```
From the code it appears that `floor_at` returns the key that is strictly smaller or `None` if there's no key smaller, and `ceil_at` returns the key which is strictly larger than `K` and `None` if there's no key larger.

It is more common to have `lower_bound` and `upper_bound`, where `lower_bound` returns the smallest key that is greater or equal, and `upper_bound` returns the smallest key that is strictly greater.

The current `ceil_at` already behaves as `upper_bound`. Can we add `upper_bound` as an alias for `ceil_at` and `lower_bound` that matches the description above?

And then expose corresponding iterators (iterate from lower_bound, iterate from upper_bound)

sergey-melnychuk replied:
ack

sergey-melnychuk replied:
done

SkidanovAlex commented on near-sdk/src/collections/tree_map.rs:
```diff
@@ -0,0 +1,1300 @@
+        } else {
+            // non-leaf node, select subtree to proceed with
+            let b = self.get_balance(r_id);
+            if (b >= 0 && lft_opt.is_some()) || rgt_opt.is_none() {
```
It should be enough to check `b >= 0` right? If we know that at least one child is set, and the left tree is no smaller than the right tree, than the left tree must exist.
Similarly, if `b < 0` rgt_opt will never be `None`

sergey-melnychuk replied:
done

SkidanovAlex commented on near-sdk/src/collections/tree_map.rs:
```diff
@@ -0,0 +1,1300 @@
+                let lft = lft_opt.unwrap();
+
+                // k - max key from left subtree
+                let k = self.max_at(lft).unwrap();
+
+                // n - id of a node that holds key k, p - id of immediate parent of n
+                let (n, p) = self.lookup_at(r_id, &k).unwrap();
```
We do two lookups here, that will traverse the exact same path, could it be merged into one lookup?

sergey-melnychuk replied:
done

SkidanovAlex commented on near-sdk/src/collections/tree_map.rs:
```diff
@@ -0,0 +1,1300 @@
+                        }
+                    };
+                } else {
+                    // n is on left link of p
```
`n` corresponds to the largest element in the subtree of `lft`. For it to be on the left link of its parent, it's parent can't be in the subtree of `lft` (otherwise it's parent is larger than `n` but `n` is the largest). Thus, the parent of `n` is `r_id` itself. Thus, below it must be that `p == r_id`. As such, the `self.lft.insert(p, l)` below is a no-op, and should be removed.

SkidanovAlex commented on near-sdk/src/collections/tree_map.rs:
```diff
@@ -0,0 +1,1300 @@
+                self.check_balance(root, &k)
+
+            } else {
+                root
```
This must be unreachable, right?
And then the previous `else if` can become `else`

sergey-melnychuk replied:
done

SkidanovAlex commented:
I didn't check all the tests, but we definitely need a test that runs many random operations (inserts, deletes) against the AVL tree, and the vanilla Rust BTree, and then checks that the contents of the two trees match, and so does the order of them.

Also, if I understand the code correctly, when you insert you assume that all the existing node ordinals are contiguous (you use `len` as the next ID, presuming that it's available). However, in the code of `do_remove` it doesn't appear that you actually make any effort not to create a hole in the ids. You replace the key of `r_id` with the key of `n`, but `n` is not the largest ID, it's the ID of one of the two adjacent elements in the tree. Probably need to make `do_remove` return the ID of the node removed, and then swap it with the `len - 1`.

sergey-melnychuk commented on near-sdk/src/collections/tree_map.rs:
```diff
@@ -0,0 +1,1300 @@
+        let k_prefix = append(&id, b'k');
+        let v_prefix = append(&id, b'v');
+
+        let root: u64 = env::storage_read(&id)
```
done

sergey-melnychuk commented on near-sdk/src/collections/mod.rs:
```diff
@@ -58,3 +68,27 @@ pub(crate) fn next_trie_id() -> Vec<u8> {
+}
+
+pub(crate) fn append_slice(id: &[u8], extra: &[u8]) -> Vec<u8> {
+    let mut result = Vec::with_capacity(id.len() + extra.len());
```
done

sergey-melnychuk commented:
@SkidanovAlex , thank you for review.

> (you use len as the next ID, presuming that it's available)

This is interesting catch, I'll try to come up with minimum test to reproduce, and make some property-based tests against `RBTree` or `Map` + sorting.

sergey-melnychuk commented:
Approach to address "use len as the next ID":

1. Increasing only `ops: u64` counter as a source of new ID for inserted node. Is only incremented when node is inserted. IMHO, not the most elegant, but with least housekeeping overhead. Problem is possible overflow when `ops` is getting close to `std::u64::MAX`. 

I doubt that such limit can ever be reached in practice (`std::u64::MAX` ~= 10^18): quick math shows that inserting an item *every nanosecond* starting now avoids overflow for next 500+ years. Yet this approach still brings a limit to keep in mind.

2. After node `n` is removed, swap node `len-1` (containing highest id) with `n`, as suggested by @SkidanovAlex . This would require updating at most 4 links: parents left or right, left, right and key. Yet to efficiently perform such update, a reverse cache `rev: Map<u64, K>` is required to avoid `O(N)` lookup of a key for node `len-1`.

This would still fit in required O(N) space and done properly allows not to keep in mind any limits.

I'll go with option 2.

sergey-melnychuk commented:
@SkidanovAlex , PTAL, all issues are resolved.

nearmax commented:
Thank you @sergey-melnychuk ! We will review it soon.

SkidanovAlex commented on near-sdk/src/collections/tree_map.rs:
```diff
@@ -0,0 +1,1637 @@
+    fn insert_n_random() {
+        test_env::setup();
+
+        for k in 1..5 {
```
5 => 500? 5 is just not serious :)

sergey-melnychuk replied:
I can't agree more! Unfortunately, even with `1..6` results in `HostError(GasLimitExceeded)`. Any suggestions how to address it?

nearmax replied:
When we create `MockedBlockchain` in `test_env::setup()` we pass `Default::default()` as a `VMConfig`. In this specific test instead of calling `test_env::setup();` we can unwrap that code and initialize `VMConfig` with `limit_config::max_gas_burnt` equal to `u64::MAX`.

Also, @SkidanovAlex `5` is not the number of items we insert in the tree here, `let n = 1 << k;` is.

SkidanovAlex commented:
- Do we have a fuzzy test that executes many random operations and compares TreeMap against BTreeMap (including order)?

- Also, given BTreeMap has `Range`, would be great to test some random ranges in such a test as well.

sergey-melnychuk commented:
> * Do we have a fuzzy test that executes many random operations and compares TreeMap against BTreeMap (including order)?
> * Also, given BTreeMap has `Range`, would be great to test some random ranges in such a test as well.

Property-based test against `std::collection::BTreeMap`:
https://github.com/near/near-sdk-rs/pull/154/files#diff-5cc570a79aec0b36a7062792ee6b3069R1605

~~Sky~~ `HostError(GasLimitExceeded)` is the limit :)

sergey-melnychuk commented:
Update:
- added method to setup test env with unlimited gas
- increased upper bound of tree size in `test_remove_n_random` to 1024
- added property-based test for range queries against BTreeMap
  - covered all 4 combination on inclusive/exclusive range bounds

Tests pass, tree seems to work properly :) Please let me know where I can improve it more!

nearmax commented:
@mikhailOK is currently reviewing it.

mikhailOK commented on near-sdk/src/collections/tree_map.rs:
```diff
@@ -0,0 +1,1681 @@
+    //
+
+    fn set_root(&mut self, root: u64) {
+        env::storage_write(&self.tree_prefix, &serialize(&root));
+        self.root = root;
```
we shouldn't have to `storage_write`: `Vector` doesn't do it with `len` field. CC @nearmax

sergey-melnychuk replied:
`root` is rather important property of a tree, not saving it would require special kind of lookup (most likely in O(N)) to restore it for an existing tree.

mikhailOK replied:
It's a bit confusing:
- simple classes don't have to handle their own storage they are stored by the owner (some key determined by the owner -> serialized object)
- near_sdk collections manage their storage and their serialization doesn't include elements
- they still have fields like `len` in Vector, and rely on those fields to be stored/read by the owner serializing the collection: `Vector::new(id)` sets `len=0`, so it can only be used for creating a new vector as opposed to providing access to an existing collection prefixed by `id`
- `TreeMap::new` calls `UnorderedMap::new` which in turn calls `Vector::new`. So we can't use `TreeMap::new` to access an existing map: even with reading `root` the `len` field is broken.

sergey-melnychuk replied:
makes sense, removing store/load of root id

sergey-melnychuk replied:
done

sergey-melnychuk commented on near-sdk/src/collections/tree_map.rs:
```diff
@@ -0,0 +1,1681 @@
+            return self.root;
+        }
+
+        let k_opt = self.key.get(&remove);
+        if k_opt.is_none() {
+            return self.root;
+        }
+        let key = k_opt.unwrap();
```
Suggested code won't compile. If there is no key to remove, no call to `do_remove` (and thus to `swap_with_last`) is made - see `remove` impl.

sergey-melnychuk replied:
Oh, I see now - applying the suggestion.

sergey-melnychuk replied:
done

sergey-melnychuk commented on near-sdk/src/collections/tree_map.rs:
```diff
@@ -0,0 +1,1681 @@
+                Some(x) => {
+                    map.insert(&target, &x);
+                },
+                None => ()
```
In fact we do, thanks.

mikhailOK replied:
How come the tests didn't catch that?
Does `prop_avl_vs_rb` catch it if we do deletions between insertions?

sergey-melnychuk replied:
Good question, seems like this is rather rare case that requires very specific input to be caught. I will try to come up with such test that must fail without `map.remove(&target)`.

mikhailOK replied:
We should improve the random test until it can catch this type of bugs:
- how does quickcheck decide the range of values? if it uses whole u32 then all insertions are distinct and deletions are no-op, need meaningful values
- how does it decide the ratio of insertions of deletions? with too many deletions the map is always empty, with almost all insertions some other case may be missed
- makes sense to test lookups between those operations

sergey-melnychuk replied:
Now I see that that test do fine, there is just no bug to catch: `remove` is unnecessary - as `target`'s height and links (max 1 link) are reset in `do_remove` before `swap_with_last` is called.

Adding comment. Sorry for confusion.

sergey-melnychuk commented on near-sdk/src/collections/tree_map.rs:
```diff
@@ -0,0 +1,1681 @@
+/// - `lookup`/`insert`/`remove`: O(log(N)) worst case
+/// - `min`/`max`: O(log(N)) worst case
+/// - `floor`/`ceil` (find closes key above/below): O(log(N)) worst case
+/// - iterate K elements in sorted order: O(Klog(N)) worst case
```
Please elaborate, I assume O(log(N)) is the lookup of the `min`. How exactly AVL-tree can be traversed in sorted order in O(K)?

sergey-melnychuk replied:
Answering my own question: with `nodes: Vector<Node<K>>` (as suggested in comment above), such iteration can be done in exactly O(log(N) + K), as `nodes` will always be sorted.

Yet insertions will degrade to O(N) worst case.

mikhailOK replied:
lookup the `min` and simply traverse the edges. Since we don't have parent pointer, we can store the stack of parents.
O(K + log(N)) nodes visited

sergey-melnychuk replied:
I don't know why I got confused - this would be plain in-order traversal, O(log(N) + K) is doable

sergey-melnychuk replied:
While it is possible to iterate over full tree in O(N) with in-order traversal, I don't think it holds for any range of K elements. For a range, finding successor node remains O(log(N)) worst case, example: range of [max(root.left).value, root.value, min(root.right).value].

[wiki](https://en.wikipedia.org/wiki/AVL_tree#Traversal)

Leaving iterator as it is for now with O(Klog(N)) worst case. In separate PR I can introduce full tree asc/desc iterators that have O(N) complexity in worst case.

sergey-melnychuk commented on near-sdk/src/collections/tree_map.rs:
```diff
@@ -0,0 +1,1681 @@
+
+    #[test]
+    fn prop_avl_vs_rb() {
+        fn prop(insert: Vec<(u32, u32)>, remove: Vec<u32>) -> bool {
+            let a = avl(&insert, &remove);
+            let b = rb(&insert, &remove);
+            let v1: Vec<(u32, u32)> = a.iter().collect();
+            let v2: Vec<(u32, u32)> = b.into_iter().collect();
+            v1 == v2
+        }
```
ack

sergey-melnychuk replied:
done

sergey-melnychuk commented on near-sdk/src/collections/heap.rs:
```diff
@@ -0,0 +1,477 @@
+        max
+    }
+
+    pub fn lookup(&self, value: &T) -> Option<u64> {
+        self.indices.get(value)
+    }
```
removed `lookup`, it was not used really

sergey-melnychuk commented on near-sdk/src/collections/heap.rs:
```diff
@@ -0,0 +1,477 @@
+    }
+}
+
+fn sort<T>(vec: &mut Vector<T>, indices: &mut UnorderedMap<T, u64>)
+    where
+        T: Ord + BorshSerialize + BorshDeserialize
+{
+    // Sort vector in-place in O(Nlog(N))
+    let n = vec.len();
+    let mut k = n;
+    while k > 1 {
+        swap(vec, 1, k, indices);
+        k -= 1;
+        sink(vec, 1, k, indices);
+    }
+}
```
replaced `sort` (which IMHO must not be included in public API of max heap) with `into_iter`

sergey-melnychuk commented on near-sdk/src/collections/tree_map.rs:
```diff
@@ -0,0 +1,1681 @@
+
+    len: u64,
+    root: u64,                      // ID of a root node of the tree
+    ht: UnorderedMap<u64, u64>,     // height of a subtree at a node
+    lft: UnorderedMap<u64, u64>,    // left link of a node
+    rgt: UnorderedMap<u64, u64>,    // right link of a node
+    key: UnorderedMap<u64, K>,      // key stored in a node
```
in such case, `nodes: Vector<Node<K>>` instead of 4 maps looks more suitable - potentially, number of lookups per node can be reduced by 4.

mikhailOK replied:
near_sdk Vector is pretty much an UnorderedMap with indices as keys, there is no resizing

sergey-melnychuk replied:
done

sergey-melnychuk commented on near-sdk/src/collections/tree_map.rs:
```diff
@@ -0,0 +1,1681 @@
+        self.max_at(self.root, self.root).map(|(k, _, _)| k)
+    }
+
+    pub fn floor(&self, key: &K) -> Option<K> {
+        self.floor_at(self.root, key)
+    }
```
ack

sergey-melnychuk replied:
done

nearmax commented:
@mikhailOK Is this collection good to go, can we merge it?

mikhailOK commented on near-sdk/src/collections/tree_map.rs:
```diff
@@ -0,0 +1,1772 @@
+        }
+    }
+
+    /// Return height of the tree - number of nodes on the longest path starting from the root node.
+    pub fn height(&self) -> u64 {
+        self.node(self.root).map(|n| n.ht).unwrap_or_default()
+    }
```
This should be private I think in case we want to switch from AVL

sergey-melnychuk replied:
done

mikhailOK commented on near-sdk/src/collections/tree_map.rs:
```diff
@@ -0,0 +1,1772 @@
+    }
+
+    /// Returns the smallest key that is greater or equal to key given as the parameter
+    pub fn floor(&self, key: &K) -> Option<K> {
```
this is still backwards compared to java TreeMap `floorKey`, `ceilingKey`.
Also one of the doc comments should be `less or equal`

sergey-melnychuk replied:
[Java](https://docs.oracle.com/javase/8/docs/api/java/util/TreeMap.html)

```
ceilingKey(K key)
Returns the least key greater than or equal to the given key, or null if there is no such key.

floorKey(K key)
Returns the greatest key less than or equal to the given key, or null if there is no such key.
```

Updating existing to:

```
/// Returns the smallest key that is greater or equal to key given as the parameter
    pub fn ceil_key(&self, key: &K) -> Option<K>

/// Returns the largest key that is less or equal to key given as the parameter
    pub fn floor_key(&self, key: &K) -> Option<K>
```

mikhailOK commented on near-sdk/src/collections/tree_map.rs:
```diff
@@ -0,0 +1,1772 @@
+    }
+
+    /// Returns the smallest key that is strictly greater than key given as the parameter
+    pub fn above(&self, key: &K) -> Option<K> {
```
Maybe `above,below` -> `higher,lower`?

sergey-melnychuk replied:
done

mikhailOK commented on near-sdk/src/collections/tree_map.rs:
```diff
@@ -0,0 +1,1772 @@
+    fn prop_avl_height() {
+        test_env::setup_free();
+
+        fn prop(insert: Vec<(u32, u32)>, remove: Vec<u32>) -> bool {
+            let map = avl(&insert, &remove);
+            map.height() <= max_tree_height(map.len())
+        }
```
I add this check to this test:
```    
    fn check_all_balances<K, V>(map: &TreeMap<K, V>, root: u64)
    where
        K: Ord + Copy + BorshSerialize + BorshDeserialize,
        V: Copy + BorshSerialize + BorshDeserialize,
    {
        if let Some(node) = map.node(root) {
            let balance = map.get_balance(&node);
            assert!(-1 <= balance && balance <= 1, "BAD BALANCE: balance {}, node id={}, ht={}, l={:?}, r={:?}", balance, node.id, node.ht, node.lft, node.rgt);
            if let Some(id) = node.lft {
                map.check_all_balances(id);
            }
            if let Some(id) = node.rgt {
                map.check_all_balances(id);
            }
        }
    }

    #[test]
    fn prop_avl_height() {
        test_env::setup_free();

        fn prop(insert: Vec<(u32, u32)>, remove: Vec<u32>) -> bool {
            let map = avl(&insert, &remove);
            check_all_balances(&map, map.root);
            map.height() <= max_tree_height(map.len())
        }

        QuickCheck::new()
            .tests(300)
            .quickcheck(prop as fn(std::vec::Vec<(u32, u32)>, std::vec::Vec<u32>) -> bool);
    }
```
and it fails, is my assert wrong?

sergey-melnychuk replied:
looking into it

sergey-melnychuk replied:
Assertion `-1 <= balance <= 1` describes perfectly balanced tree (at each level), height of such tree would always be `log(N+1)`. So assertion is not wrong, just too strict for AVL tree.

AVL gives no guarantees of being perfectly balanced at each level after each insert/remove, thus `log(N+1)` is a lower bound of tree height, and upper bound is: `C * log2(n + D) + B` according to [wiki](https://en.wikipedia.org/wiki/AVL_tree).

mikhailOK replied:
Quote from wiki:
```
of its two child sub-trees. A binary tree is defined to be an AVL tree if the invariant

    BalanceFactor ( n o d e ) ∈ { − 1 , 0 , 1 } {\displaystyle {\text{BalanceFactor}}(node)\in {\{-1,0,1\}}} {\displaystyle {\text{BalanceFactor}}(node)\in {\{-1,0,1\}}}[7]

holds for every n o d e {\displaystyle node} {\displaystyle node} in the tree. 
```

sergey-melnychuk replied:
still looking into it then

sergey-melnychuk replied:
`remove` indeed does not restore AVL-tree invariants, investigating further

sergey-melnychuk replied:
Thank you, @mikhailOK ! I was wrong relying only on tree height to validate balance, as well as about definition of a perfectly balanced tree (it is only perfectly balanced when balance == 0 in all nodes), and also about AVL-tree fundamental property - balance among all subtrees.

Fixed now, `prop_avl_balance` passes.

mikhailOK commented on near-sdk/src/collections/tree_map.rs:
```diff
@@ -0,0 +1,1776 @@
+
+impl<K, V> TreeMap<K, V>
+    where
+        K: Ord + Copy + BorshSerialize + BorshDeserialize,
+        V: Copy + BorshSerialize + BorshDeserialize,
```
I think `V` doesn't need `Copy` at all
And for `K` it makes sense to remove `Copy` and require `Clone` or change some interfaces to `&K`

sergey-melnychuk replied:
ack

sergey-melnychuk replied:
PTAL, now `K: Clone + Ord + <ser/de>` and `V: <ser/de>` only.
I was trying to keep `tree` as `Vector<Node<&K>>`, but lost this fight to borrow checker, thus put a few `.clone()`s when moving keys around.

nearmax commented:
@mikhailOK said he found a test that surfaces some bug in the implementation. @mikhailOK could you please provide it here?

nearmax commented:
@SkidanovAlex @mikhailOK LGTMed this PR. Is there anything from your side you want to comment on?

SkidanovAlex commented:
The AVL tree looks good to me with all the new tests.

Can we remove Heap and the HeapMap? AVL tree is a strict superset of functionality.

Let's also get LGTM from Mikhail.

nearmax commented:
> The AVL tree looks good to me with all the new tests.
> 
> Can we remove Heap and the HeapMap? AVL tree is a strict superset of functionality.
> 
> Let's also get LGTM from Mikhail.

AFAIU heap has smaller constant in O(log N) than AVL, so it is beneficial to keep around. @mikhailOK already LGTMed, see above.

mikhailOK commented:
Heap is kind of weird right now, it's a priority queue but that also doesn't allow duplicate values. Make sense to remove it or make private until we decide what we want from it.

nearmax commented:
@sergey-melnychuk Could you remove the heap? After that we will merge and I will release the bounty.

sergey-melnychuk commented:
@nearmax , heap is removed - I'll submit separate PR with priority queue, then you can decide if it is useful for you.

nearmax commented:
@SkidanovAlex could you change your review to "approved" so that I don't force merge?

nearmax commented:
@sergey-melnychuk Thank you very much for spending your time on implementing this collection together with us! I have paid out the bounty, please let us know if there any issues with payments. We are very satisfied with the result and hoping for the future collaboration together with you.

sergey-melnychuk commented:
@nearmax , Thank you, nice to know! Please feel free to ping me directly on GitHub if you need my help. 

As promised, submitting a PR with Priority Queue: https://github.com/near/near-sdk-rs/pull/170
